File: src/main/java/com/teamwizardry/wizardry/common/spell/effect/EffectArcane.java
package com.teamwizardry.wizardry.common.spell.effect;

import com.teamwizardry.wizardry.Wizardry;
import com.teamwizardry.wizardry.api.spell.Instance;
import com.teamwizardry.wizardry.api.spell.Interactor;
import com.teamwizardry.wizardry.api.spell.PatternEffect;
import com.teamwizardry.wizardry.api.utils.RandUtil;
import com.teamwizardry.wizardry.client.particle.GlitterBox;
import com.teamwizardry.wizardry.common.init.ModSounds;
import net.minecraft.util.DamageSource;
import net.minecraft.world.World;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;

import java.awt.*;

/*
* By: Carbon
* Pure magic damage effect.
* TODO: Particles
* */
public class EffectArcane extends PatternEffect {
    // FIXME: Needs balancing.
    public static final float POTENCY_MULTIPLIER = 1f;

    @Override
    public void affectEntity(World world, Interactor entity, Instance instance) {
        if(entity.getType() != Interactor.InteractorType.ENTITY) return;

        entity.getEntity().attackEntityFrom(DamageSource.MAGIC, (float) instance.getAttributeValue("potency") * POTENCY_MULTIPLIER);
        ModSounds.playSound(world, instance.getCaster(), entity, ModSounds.FIREWORK, 0.1f);
    }

    // No affect on blocks. This remains a NO-OP.
    @Override
    public void affectBlock(World world, Interactor block, Instance instance) {
        // NO-OP
    }

    private static final Color[] colors = new Color[]{Color.CYAN, Color.BLUE, Color.MAGENTA};

    //  Graphical Methods

    @Override
    public Color[] getColors() {
        return colors;
    }

    @OnlyIn(Dist.CLIENT)
    @Override
    public void runClient(World world, Instance instance, Interactor target) {
        for (int i = 0; i < 100; i++)
            Wizardry.PROXY.spawnParticle(
                    new GlitterBox.GlitterBoxFactory()
                            .setOrigin(target.getPos()
                                    .add(RandUtil.nextDouble(-0.15, 0.15),
                                            RandUtil.nextDouble(-0.15, 0.15),
                                            RandUtil.nextDouble(-0.15, 0.15)))
                            .setTarget(RandUtil.nextDouble(-0.5, 0.5),
                                    RandUtil.nextDouble(-0.5, 0.5),
                                    RandUtil.nextDouble(-0.5, 0.5))
                            .setDrag(RandUtil.nextFloat(0.2f, 0.3f))
                            .setGravity(RandUtil.nextFloat(-0.005f, -0.015f))
                            .setInitialColor(getRandomColor())
                            .setGoalColor(getRandomColor())
                            .setInitialSize(RandUtil.nextFloat(0.1f, 0.3f))
                            .setGoalSize(0)
                            .setInitialAlpha(RandUtil.nextFloat(0.5f, 1))
                            .createGlitterBox(RandUtil.nextInt(5, 25)));

    }
}

File: src/main/java/com/teamwizardry/wizardry/common/spell/effect/EffectArcane.kt
package com.teamwizardry.wizardry.common.spell.effect

import com.teamwizardry.wizardry.common.spell.component.Attributes
import com.teamwizardry.wizardry.common.spell.component.Instance
import net.fabricmc.api.Environment
import java.awt.Color

/*
* By: Carbon
* Pure magic damage effect.
* */
class EffectArcane : PatternEffect() {
    fun affectEntity(world: World?, entity: Interactor, instance: Instance) {
        if (entity.getType() != InteractorType.ENTITY) return
        entity.getEntity()
            .damage(DamageSource.MAGIC, instance.getAttributeValue(Attributes.INTENSITY).toFloat() * POTENCY_MULTIPLIER)
        playSound(world, instance.caster, entity, ModSounds.FIREWORK, 0.1f)
    }

    // No affect on blocks. This remains a NO-OP.
    fun affectBlock(world: World?, block: Interactor?, instance: Instance?) {
        // NO-OP
    }

    //  Graphical Methods
    val colors: Array<Color>
        get() = Companion.colors

    @Environment(EnvType.CLIENT)
    fun runClient(world: World?, instance: Instance?, target: Interactor) {
        for (i in 0..99) Wizardry.Companion.PROXY.spawnParticle(
            GlitterBoxFactory()
                .setOrigin(
                    target.getPos()
                        .add(
                            RandUtil.nextDouble(-0.15, 0.15),
                            RandUtil.nextDouble(-0.15, 0.15),
                            RandUtil.nextDouble(-0.15, 0.15)
                        )
                )
                .setTarget(
                    RandUtil.nextDouble(-0.5, 0.5),
                    RandUtil.nextDouble(-0.5, 0.5),
                    RandUtil.nextDouble(-0.5, 0.5)
                )
                .setDrag(RandUtil.nextFloat(0.2f, 0.3f))
                .setGravity(RandUtil.nextFloat(-0.005f, -0.015f))
                .setInitialColor(getRandomColor())
                .setGoalColor(getRandomColor())
                .setInitialSize(RandUtil.nextFloat(0.1f, 0.3f))
                .setGoalSize(0f)
                .setInitialAlpha(RandUtil.nextFloat(0.5f, 1f))
                .createGlitterBox(RandUtil.nextInt(5, 25))
        )
    }

    companion object {
        // FIXME: Needs balancing.
        const val POTENCY_MULTIPLIER = 1f
        private val colors = arrayOf(Color.CYAN, Color.BLUE, Color.MAGENTA)
    }
}